app/controllers/mixins/ems_common/angular.rb

```diff
@@ -354,6 +363,10 @@ def ems_form_fields
           host_default_vnc_port_end = @ems.host_default_vnc_port_end.to_s
         end
 
+        if @ems.kind_of?(ManageIQ::Providers::Amazon::CloudManager)
```

Comment:
In the interest of plugability, we want to avoid constructs like the above as much as possible.

This should be replaced by `@ems.supports? ...`

Reply:
@agrare do you agree with @mzazrivec ? Should I patch `ExtManagementSystem` for that?
@mzazrivec considering that this method all consists of such conditions - do you want me to fix everything, or just that? Is it ok that I'll introduce more mess here?

Reply:
P.S. I understand that we should refactor such spaghetti code. But such occasional refactoring using patches on patches - we just get more spaghetti. So, I believe, it should be a separate task with a complex and weighted approach, with discussions and decisions. Please don't force me to make scope creep here.

Reply:
P.P.S. For a proper refactoring we can use something like "**Strategy**" / "**Decorator**" pattern, and/or something like this: http://trailblazer.to/gems/cells.

Reply:
@mzazrivec I completely agree that we need to move in that direction, however it is a little impractical to expect that kind of refactoring PR by PR in my opinion.  What @AlexanderZagaynov did here fits with the current pattern in this class.

Reply:
Well. For several years already the "provider plugability" project was concluded by Marcel. Since that time it's no reason someone should be adding such code.

Especially if we allow the core team members do these kinds of things we have no arguments to want people from outside or other teams play according to the rules.

Therefor I strongly agree with with @mzazrivec. 

Actually I keep repeating this argument over and over and people keep with coming with excuses why "this time" it needs to be done. 

I already said at some other place that I am no longer merging these. If someone merges it, well Ok, I have to live and deal with that. By I am not merging it.

Reply:
@martinpovolny if so, can you please answer to my questions here: https://github.com/ManageIQ/manageiq-ui-classic/pull/5621#discussion_r288024052 and here: https://github.com/ManageIQ/manageiq-ui-classic/pull/5621#discussion_r288033900 please?

Reply:
@martinpovolny we should refactor this method to match how we want it to be.  People just copy what they see if they don't understand it, so lets show them how we want it to look.

Trying to do this just with `.supports?()` looks like it might be unwieldy, would allowing the providers to define their own dialog help out here?  Thinking we can have the defaults for Infra/Cloud then allow the providers to subclass and add specifics.

Because just adding a `.supports?(:assume_role)` is really no different from `ems.kind_of?(Amazon)` if amazon is the only provider that will ever have this concept.

Reply:
> we should refactor this method to match how we want it to be. People just copy what they see if they don't understand it, so lets show them how we want it to look.

:heart: I love to have more people who understand that we actually NEED to do a lot of refactoring on the UI. We have hundreds of places that need this kind of work. We need resources and management approval for that work. Thank you!

> Because just adding a .supports?(:assume_role) is really no different from ems.kind_of?(Amazon) if amazon is the only provider that will ever have this concept.

It actually is quite different. Because when reading the code you are going to know that what comes next is related to "assume_role". Which is much more that knowing that it's related to "Amazon". It's a one less step for whoever is going to try to refactor this piece of code in the future.

> Trying to do this just with .supports?() looks like it might be unwieldy, would allowing the providers to define their own dialog help out here? Thinking we can have the defaults for Infra/Cloud then allow the providers to subclass and add specifics.

We are going to rewrite these forms to DDF and use the API. Each capability that you create and `.support()` call that you add will help the poor fellow who is going to figure what these lines mean and what is the form supposed to do when rewriting it.


￼

Reply:
> It actually is quite different. Because when reading the code you are going to know that what comes next is related to "assume_role". Which is much more that knowing that it's related to "Amazon". It's a one less step for whoever is going to try to refactor this piece of code in the future.

That's a fair point it is more descriptive, I still think that trying to build a dialog in one method for every provider based on a bunch of `if/elsif/else supports?()` conditions will be _almost_ as bad and this is something that should be solved by subclassing.

Reply:
@agrare because of those "ifs" I proposed https://en.wikipedia.org/wiki/Strategy_pattern
I guess we should move part of UI-related code from UI repo to a providers (and core for a basic class). It can be done with that "cells" gem I mentioned above. Or we can just make decorators. Like, basic "edit" form, and specific decorators for providers and auth types. Anyway, it will help us to avoid those ifs. Currently, for example, providers data collectors/parsers work same way.

Reply:
And instead of `supports?(:something)` I guess better to use something like `ems.give_me_fields_you_need_and_how_to_format_them`.

Reply:
> And instead of supports?(:something) I guess better to use something like ems.give_me_fields_you_need_and_how_to_format_them().

Yeah I agree I think this is what the data-driven forms would be giving us.

Reply:
The pattern that we need here is: translate the list of capabilities of the provider into a list of capabilities for the UI that is going to be displayed. Need authorization or kind a-b-c? Add it. Need multiple endpoints, add it. Etc. Capabilities not provider names.

I know that the provider forms do not follow this pattern and for various reasons it was not in my power to fix that or prevent this from happening.

However @mzazrivec, @himdel, @h-kataria, myself, all reviewers on the UI are trying to ensure that with each PR the situation gets a bit better not worse.

Until a point when we have the time, API and mandate to actually rewrite the particular form, we have to carry on like this.

So please, when adding new stuff. Take the time, think about what is the the capability that you are adding. Name it correctly. Query it.

Other that that. Leave the mess as it is. We will need to rewrite this. Just do not make the rewrite more expensive than needed.

Reply:
done